### PR TITLE
ARROW-12496: [C++][Dataset] Ensure AsyncScanner is covered by all scanner tests

### PR DESCRIPTION
**Pull request by lidavidm**
Some tests directly constructed a Scanner from a Fragment, which isn't supported by the async scanner. This introduces a simple dataset wrapper so all these tests will actually test both sync/async scanners.

### CODE REVIEW

File: cpp/src/arrow/dataset/scanner_test.cc [hunk 1]
```diff
@@ -71,6 +71,14 @@ std::ostream& operator<<(std::ostream& out, const TestScannerParams& params) {
 
 class TestScanner : public DatasetFixtureMixinWithParam<TestScannerParams> {
  protected:
+  std::shared_ptr<Scanner> MakeScanner(std::shared_ptr<Dataset> dataset) {
+    ScannerBuilder builder(dataset, options_);
```
Comment (westonpace) [inline cpp/src/arrow/dataset/scanner_test.cc:75]: Nit: could be `std::move(dataset)`

File: cpp/src/arrow/dataset/scanner_test.cc [hunk 2]
```diff
@@ -319,6 +322,23 @@ class FailingFragment : public InMemoryFragment {
...
+      Future<std::shared_ptr<RecordBatch>> operator()() {
+        if (index > 16) {
+          return Status::Invalid("Oh no, we failed!");
+        }
+        auto batch = batches[index++ % batches.size()];
+        return Future<std::shared_ptr<RecordBatch>>::MakeFinished(batch);
+      }
+      RecordBatchVector batches;
+      int index = 0;
+    } Generator;
```
Comment (westonpace) [inline cpp/src/arrow/dataset/scanner_test.cc:338]: Ah, this is a neat trick for me to learn.  Is this basically creating an instance of an anonymous struct?

**Review by westonpace — APPROVED** [review]
Thanks!  I appreciate the extra testing.  I had noticed that some tests create a scanner from a fragment and didn't want to introduce that complexity into `AsyncScanner` for the sake of tests.  This is a clever workaround.

File: cpp/src/arrow/dataset/scanner_test.cc (hunk 2 above)
Comment (lidavidm) [reply cpp/src/arrow/dataset/scanner_test.cc:338]: Yup, it's used quite a bit in the datasets implementation (e.g. see ParquetScanTask in file_parquet.cc).

**Comment by lidavidm** [comment]
Tests pass minus the known flaky ones (Ruby segfaulting, JNI build having issues)

**Pull request closed without merging**